modules/rostests/kmtests/kmtest/service.c

```diff
@@ -101,6 +102,102 @@ KmtCreateService(
                              ServiceHandle);
 }
 
+/**
+ * @name KmtGetServiceStateAsString
+ *
+ * @param ServiceState
+ *        Service state as a number
+ *
+ * @return Service state as a string
+ */
+static
+PCSTR
+KmtGetServiceStateAsString(
+    IN DWORD ServiceState)
+{
+    switch(ServiceState)
+    {
+        case SERVICE_STOPPED:
+            return "STOPPED";
+        case SERVICE_START_PENDING:
+            return "START_PENDING";
+        case SERVICE_STOP_PENDING:
+            return "STOP_PENDING";
+        case SERVICE_RUNNING:
+            return "RUNNING";
+        case SERVICE_CONTINUE_PENDING:
+            return "CONTINUE_PENDING";
+        case SERVICE_PAUSE_PENDING:
+            return "PAUSE_PENDING";
+        case SERVICE_PAUSED:
+            return "PAUSED";
+        default:
+            return "(Unknown)";
+    }
+}
+
+/**
+ * @name KmtCheckServiceState
+ *
+ * @param ServiceName
+ *        Name of the service to check,
+ *        or NULL
+ * @param ServiceHandle
+ *        Handle to the service
+ * @param ExpectedServiceState
+ *        State which the service should be in
+ *
+ * @return Win32 error code
+ */
+static
+DWORD
+KmtCheckServiceState(
+    IN PCWSTR ServiceName OPTIONAL,
+    IN SC_HANDLE ServiceHandle,
+    IN DWORD ExpectedServiceState)
+{
+    DWORD Error = ERROR_SUCCESS;
+    SERVICE_STATUS ServiceStatus;
+    DWORD StartTime = GetTickCount();
+    DWORD Timeout = 10 * 1000;
+    PCWSTR ServiceNameOut = ServiceName ? ServiceName : L"(handle only, no name)";
+
+    assert(ServiceHandle);
+    assert(ExpectedServiceState);
+
+    if (!QueryServiceStatus(ServiceHandle, &ServiceStatus))
+        error_goto(Error, cleanup);
+
+    while (ServiceStatus.dwCurrentState != ExpectedServiceState)
+    {
+        // NB: Not bothering with ServiceStatus.dwWaitHint nor ServiceStatus.dwCheckPoint.
+
+        Sleep(1 * 1000);
+
+        if (!QueryServiceStatus(ServiceHandle, &ServiceStatus))
+            error_goto(Error, cleanup);
+
+        if (GetTickCount() - StartTime >= Timeout)
+            break;
+    }
+
+    if (ServiceStatus.dwCurrentState != ExpectedServiceState)
+    {
+        ok(FALSE, "Service = %ls, state = %lu %s (!= %lu %s)\n",
+           ServiceNameOut,
+           ServiceStatus.dwCurrentState, KmtGetServiceStateAsString(ServiceStatus.dwCurrentState),
+           ExpectedServiceState, KmtGetServiceStateAsString(ExpectedServiceState));
+        goto cleanup;
```

Comment:
Wouldn't you want to return failure in this case?

Reply:
I could if you prefer, though my reasoning was: fail for subroutine unexpected error only.
Here, we succeed at checking what we want. Only the result is false, which we do report as a test failure.
In other word, it is a minimal change of behavior, that can later be made harder if need be.
This is also kind of why I initially called it "Check" only.